### ISSUE TITLE
The search()/pulldata() handler consumes files meant to be used as external secondary instances

### ISSUE THREAD

**Issue by lognaturel** [getodk/collect#3335]
To reproduce this issue, #3334 first needs to be addressed.

Then, try a form like [this one](https://docs.google.com/spreadsheets/d/1xZDS2XqwbMcgqa5nvi5_Ug03WkijogcQh4yPNexqaE8/edit#gid=0) that attempts to use a CSV file as an external secondary instance. It works the first time the form is opened but fails the second time because the CSV has been imported to a db by the search()/pulldata() machinery and renamed.

Some options I see to address this:
- instead of renaming CSVs to mark them as ingested, leave the files untouched and use a file hash to detect a change
- instead of greedily putting all CSVs in a DB, only read in the ones used for search()/pulldata() (or not used as external secondary instance)

**Comment by breakbusyloop** [comment]
I barely have a sense of what's going on here, but I see enough to have a strong opinion. Using a secondary characteristic (file name change) as a source of primary information has the same pitfalls as the use of sentinels, and will forever be a bug factory.

I really prefer the first option:
 * instead of renaming CSVs to mark them as ingested, leave the files untouched and use a file hash to detect a change

**Comment by lognaturel** [comment]
> Using a secondary characteristic (file name change) as a source of primary information has the same pitfalls as the use of sentinels, and will forever be a bug factory.

➕➕➕

I agree that the first option would be much better.

You're getting into some really head-scratching territory here. You'll see that search()/pulldata() is a different way of accessing data in files external to the form definition as what is described in #3334.

https://forum.opendatakit.org/t/external-data-current-state-of-affairs should hopefully provide some context.

**Comment by breakbusyloop** [comment]
It looks like the CSV file renaming happens in ExternalDataReaderImpl.doImport. The file will be reimported if it shows up again with the same name. After we stop renaming files, we could reimport based on a file hash (which requires reading the whole file, right?), or we could reimport based on simpler attributes like the file timestamp.

**Comment by lognaturel** [comment]
Yes, the hash does require reading through the whole file. See `FileUtils.getMd5Hash` for how that's currently done. I think it's ok to do that work since there isn't particularly a need for high performance when downloading a new form version. But I can't see a reason why file timestamp wouldn't be sufficient in this case.

**Comment by breakbusyloop** [comment]
I'm noticing some expected database structure in ExternalDataUtil.java. We could include the previously imported file timestamp in the database as a separate ODK metadata table. Other metadata could include the ODK version used to create the database file. Is there a broader need for such metadata, or does it already exist?

**Comment by lognaturel** [comment]
> include the previously imported file timestamp in the database as a separate ODK metadata table

That seems reasonable to me. I don't think any of this kind of metadata is currently stored.

**Comment by breakbusyloop** [comment]
It appears that the search/pulldata feature always consumes any CSV files in the form media directory regardless of whether they are referenced in the form. As a first step in this fix I'll look into having the CSV ingestion happen only for relevant files. The bug could still be present if the form references the same CSV file as an external secondary instance as well as a search/pulldata reference.

**Comment by breakbusyloop** [comment]
(notes to self and other devs following this work...)

- The search/pulldata features landed in 2014 as change 2f9c1b6, and did not include any tests.
- The example files included in that change that demonstrated this feature have since gone missing. I've attached the one included in the original change. [Sample-SearchAndSelect.zip](https://github.com/opendatakit/collect/files/3933419/Sample-SearchAndSelect.zip)
- A cautionary code comment claims importing data must be done before the form XML is loaded ("This order is important. Import data, then initialize."). That poses a challenge to my prior suggestion that we only import the files that referenced by the form.

**Comment by breakbusyloop** [comment]
@opendatakit-bot claim

**Comment by breakbusyloop** [comment]
@opendatakit-bot claim --force

**Comment by breakbusyloop** [comment]
I believe I have addressed this now in commit 62f31d6. The External Data handler for search()/pulldata() will still import any CSV file it comes across, but it will no longer rename the file. Instead it tracks the file modification timestamp in a metadata table in the SQLite database file holding the ingested data.

**Comment by lognaturel** [comment]
@OpenDataNerd That commit looks really great.

**Comment by breakbusyloop** [comment]
@ln thanks. There were many commits before it that will be in the overall PR (lots of refactoring for SQLite querying), but I'm waiting until my other PR is accepted before posting this one (which depends on it, but GitHub doesn't supported chained PRs).

**Comment by breakbusyloop** [comment]
The fix is ready but is blocked by a PR it depends on.

**Comment by breakbusyloop** [comment]
@opendatakit-bot claim

**Comment by lognaturel** [comment]
> The fix is ready but is blocked by a PR it depends on.

Hi, @OpenDataNerd! It would be really great to get this in soon. Any chance you could submit what you've already done as a PR?

**Comment by breakbusyloop** [comment]
@opendatakit-bot claim

**Comment by breakbusyloop** [comment]
@lognaturel I'll try to get to this before the weekend

**Comment by lognaturel** [comment]
I think that there is possibility for a bad state currently if the device time was accidentally ahead of what it should be, a form was downloaded, then the time was set correctly and an update to a CSV attachment was pushed to the device or downloaded after that. Then the newer version would look older and not be imported. I'm wondering whether we should just re-ingest if the timestamps don't match rather than if the file write time is after the database time. I believe that would take care of all cases. CC @seadowg since you got me thinking about timestamp vs. md5.

**Comment by seadowg** [comment]
@lognaturel yeah I think that's better. I kind of think an MD5 (or any fast hash) would be best as it just avoids time entirely.
